### DISCUSSION TITLE
Iterating an object and erasing members

### DISCUSSION THREAD

**Discussion by awolff64** [danielaparker/jsoncons#361] [Show and tell] [answered]
I assumed the object works like an std::map and learned that the container is a vector.
This cause the following code to fail

	auto it = m_Data.object_range().begin();
	while (it != m_Data.object_range().end())
	{
		if (lw::util::general::MatchText(it->name(), Key, true))
		{
			auto tmp = it;
 
			it++;
			// it now points to the successor of the to be erased element
			m_Data.erase(tmp);
			// tmp is now invalid
			continue;
		}
		it++;
	}

If I have to erase the member I copy the iterator to tmp and increment the iterator. This is to ensure that it points to a valid element after erase.
This did not work as the object's container is a vector. To get it working I do not increment the iterator when I erase. After the ease the it already points to the next object.

	auto it = m_Data.object_range().begin();
	while (it != m_Data.object_range().end())
	{
		if (lw::util::general::MatchText(it->name(), Key, true))
		{
			m_Data.erase(it);
			// it now points to the successor of the erased element
			continue;
		}
		it++;
	}

Having a solution is fine.
However I would prefer being agnostic to the underlying container. 

Is there a better way to get this working?
E.g. if erase would return the iterator to the next element (as std::vector::erase does, but not std::map::erase).

**Comment by danielaparker** [comment]
Yes, jsoncons uses flat-map semantics, currently using `std::vector` as underlying storage. This should be more prominently documented.

I think
```
m_Data.erase(it);
// it now points to the successor of the erased element
```

is actually undefined behavior, I think it should be
```
it = m_Data.erase(it);
```
except the jsoncons `erase` function returns void. We'll fix that in a patch, after which it should be written as above.

Daniel

**Comment by danielaparker** [comment, marked as answer]
This change has been made on master, the `basic_json::erase` function return value, previously void, is now an iterator following the last removed element,
```
array_iterator erase(const_array_iterator pos);
object_iterator erase(const_object_iterator pos);

array_iterator erase(const_array_iterator first, const_array_iterator last); 
object_iterator erase(const_object_iterator first, const_object_iterator last);
```

Incidentally, instead of `it->name()`, use `it->key()`, `name()` was deprecated a long time ago, that should have resulted in a warning. To be sure that you're not using any deprecated functions, you may want to define the `JSONCONS_NO_DEPRECATED` macro.
